### ISSUE TITLE
[QUESTION] 

### ISSUE THREAD

**Issue by WSUFan** [hibiken/asynq#717]
@gunnsth Thanks for opening an issue and asking this question.

You should be able to run multiple worker processes connecting to the same Redis instance and read from the same queue. Alternatively, you can run multiple worker processes with each reading from different queues.

There's  `asynqmon ps` command to list all workers processes.

Please let me know if you run into any issues!

_Originally posted by @hibiken in https://github.com/hibiken/asynq/issues/88#issuecomment-587272006_

**Comment by hibiken** [comment]
@WSUFan thanks for pointing this out. 

I believe you're right, this needs to be fixed.
Approaches I can think of:
1) take a lock on these redis keys (asynq:{<qname>}:lease) when running `listLeaseExpired` command (lock with TTL to avoid unfortunate situation)
2) within `listLeaseExpired`, we could move these expired task ids to "stagin" area so that they are not in the asynq:{<qname>}:lease anymore 

Any feedback is appreicated.

**Comment by WSUFan** [comment]
yeah, maybe a redis lock is needed when operating with these keys. I would choose approach 1, haha

**Comment by yousifh** [comment]
While there is a race condition, is there an actual risk of task duplication? 

Take 2 workers W1 and W2 and an expired task T1 and consider this timeline of sequential events

- W1 calls `ListLeaseExpired` and gets T1
- W2 calls `ListLeaseExpired` and gets T1
- W1 calls `retry` It removes T1 from `asynq:{<qname>}:active` and `asynq:{<qname>}:lease` and moves it to `asynq:{<qname>}:retry`
- W2 calls `retry` but it will fail because calling `LREM` on `asynq:{<qname>}:active` return 0 elements. 

Now of course both W1 and W2 having the same expired tasks is not ideal and there is a race condition where in the last step above, W2 might `LREM` task T1 after it has been re-enqueued again. 

So I think your proposal of a Redis lock around `listLeaseExpired` is good solution so each worker is dealing with disjointed set of expired tasks

**Comment by WSUFan** [comment]
I believe an another solution would be to let the client side handle this. We could invoke the ```listLeaseExpired()``` function on the client side. In this scenario, there would only be one retry, assuming the task ID is unique.
